gson/src/main/java/com/google/gson/Gson.java

```diff
@@ -523,46 +524,76 @@ public <T> TypeAdapter<T> getAdapter(TypeToken<T> type) {
+    LinkedHashMap<TypeToken<?>, TypeAdapter<?>> threadCalls = calls.get();
+    boolean isInitialAdapterRequest = false;
     if (threadCalls == null) {
-      threadCalls = new HashMap<>();
+      threadCalls = new LinkedHashMap<>();
       calls.set(threadCalls);
-      requiresThreadLocalCleanup = true;
+      isInitialAdapterRequest = true;
     }
 
     // the key and value type parameters always agree
     @SuppressWarnings("unchecked")
-    FutureTypeAdapter<T> ongoingCall = (FutureTypeAdapter<T>) threadCalls.get(type);
+    TypeAdapter<T> ongoingCall = (TypeAdapter<T>) threadCalls.get(type);
     if (ongoingCall != null) {
       return ongoingCall;
     }
 
+    int existingAdaptersCount = threadCalls.size();
+    boolean foundCandidate = false;
     try {
       FutureTypeAdapter<T> call = new FutureTypeAdapter<>();
       threadCalls.put(type, call);
 
       for (TypeAdapterFactory factory : factories) {
         TypeAdapter<T> candidate = factory.create(this, type);
         if (candidate != null) {
-          @SuppressWarnings("unchecked")
-          TypeAdapter<T> existingAdapter = (TypeAdapter<T>) typeTokenCache.putIfAbsent(type, candidate);
-          // If other thread concurrently added adapter prefer that one instead
-          if (existingAdapter != null) {
-            candidate = existingAdapter;
-          }
-
           call.setDelegate(candidate);
+          // Replace future adapter with actual adapter
+          threadCalls.put(type, candidate);
+
+          if (isInitialAdapterRequest) {
+            // Publish resolved adapters to all threads
+            // Can only do this for the initial request because cyclic dependency TypeA -> TypeB -> TypeA
+            // would otherwise publish adapter for TypeB which uses not yet resolved adapter for TypeA
+            // See https://github.com/google/gson/issues/625
+            for (Map.Entry<TypeToken<?>, TypeAdapter<?>> resolvedAdapterEntry : threadCalls.entrySet()) {
+              typeTokenCache.putIfAbsent(resolvedAdapterEntry.getKey(), resolvedAdapterEntry.getValue());
+            }
+
+            @SuppressWarnings("unchecked")
+            TypeAdapter<T> actualAdapter = (TypeAdapter<T>) typeTokenCache.get(type);
+            // Prefer the actual adapter, in case putIfAbsent call above had no effect because other
+            // thread already concurrently added other adapter instance for the same type
+            candidate = actualAdapter;
+          }
+          foundCandidate = true;
           return candidate;
         }
       }
       throw new IllegalArgumentException("GSON (" + GsonBuildConfig.VERSION + ") cannot handle " + type);
     } finally {
-      threadCalls.remove(type);
-
-      if (requiresThreadLocalCleanup) {
+      if (isInitialAdapterRequest) {
         calls.remove();
       }
+      if (!foundCandidate) {
+        Iterator<TypeAdapter<?>> adaptersIterator = threadCalls.values().iterator();
+        // Skip existing non-broken adapters
```

Comment:
Yes, I have now removed this logic here again because this is indeed likely a contrived corner case, and the current implementation does not handle that in a better way either.